### PR TITLE
feat(agents): add passive execution health monitoring for death spiral detection

### LINKED ISSUE #40445: feat: passive execution health monitoring for agent death spirals
## Problem

Agents can enter execution death spirals where they produce artifacts (planning files, checklists, markdown documents) instead of executing real work. This happens when:

1. The agent hits a persistent blocker (auth failure, rate limit, tool unavailability)
2. A tight SLA watchdog or heartbeat forces the agent to produce *something* every N seconds
3. The agent substitutes file creation for real execution to satisfy the activity requirement

**Real-world data:** In one observed session, an agent created 220 planning files in 4 hours after hitting an Upwork auth blocker. Zero commits, zero pushes, zero real actions. Approximately $50 in tokens burned with no productive output. The user only discovered the problem by manually reviewing session logs.

This pattern is invisible to users because OpenClaw tracks token usage and tool call counts, but does not distinguish between productive and non-productive tool calls.

## Proposed solution

A passive `ExecutionHealthMonitor` that evaluates tool call patterns after each turn and emits warnings when it detects:

| Pattern | What it detects | Default threshold |
|---------|----------------|-------------------|
| `file-burst` | Rapid file creation in a short window | 10 writes per window |
| `tool-repeat` | Same tool+args called repeatedly | 5 identical calls |
| `no-effect-loop` | Consecutive turns without real side-effects (commit, push, send) | 10 turns |
| `error-cascade` | Consecutive tool call failures | 3 errors |

Detection is passive (warning-only logging), enabled by default, with configurable thresholds via `agents.defaults.executionHealth`. Zero performance overhead since it piggybacks on existing session message data.

## Non-goals (follow-up work)

- Auto-throttle (increasing heartbeat interval during spirals)
- Circuit breaker (stopping execution after N non-productive turns)
- CLI dashboard (`openclaw status --health`)
- Context engine integration (injecting warnings into system prompt)

These are intentionally left for follow-up PRs to keep scope tight.

### PR DESCRIPTION
**Pull request by dsantoreis**
## Summary

Adds an `ExecutionHealthMonitor` that detects agent death spirals by analyzing tool call patterns after each turn. Four detection patterns catch the most common failure modes: file creation bursts, tool call loops, turns without real side-effects, and error cascades.

Closes #40445

## What changed

- New `src/agents/execution-health.ts` with `ExecutionHealthMonitor` class and 4 pattern detectors
- New `src/agents/execution-health.test.ts` with 18 tests covering all patterns, edge cases, config overrides, and state reset
- Config slot at `agents.defaults.executionHealth` with zod validation, help text, and labels
- Integration in `attempt.ts` turn loop: evaluates signals after each afterTurn hook and logs warnings

## What did NOT change

- No auto-throttle or circuit breaker (follow-up scope)
- No CLI integration (`openclaw status --health` is future work)
- No context engine hook integration (engines can be extended independently)
- Default behavior is warning-only logging; no execution is blocked or modified

## Context

Based on real-world observation: an agent created 220 planning files in 4 hours after hitting an auth blocker with a 60-second SLA watchdog. The agent substituted file creation for real execution, burning ~$50 in tokens with zero productive output. This pattern is currently invisible to users.

## Validation

```
pnpm vitest run src/agents/execution-health.test.ts   # 18/18 pass
pnpm build                                              # clean
pnpm oxfmt --check                                      # clean
```

---
AI-assisted (Claude). Fully tested and reviewed.

### CODE REVIEW

**Comment by dsantoreis** [comment]
CI failure is real (not transient), so I did not rerun blindly. Root causes from run 22843223950:\n\n1) New tests in src/agents/execution-health.test.ts are using legacy message/tool payload shapes that no longer match AgentMessage/ToolCall contracts (missing timestamp, wrong content union, missing arguments).\n2) src/cli/gateway-cli/run-loop.test.ts mocks Record<Signals,...> with only 3 keys, but the type now requires full Signals coverage.\n3) detect-secrets flagged appcast.xml:77 as Base64 high-entropy (likely false positive, needs explicit allowlist handling or content adjustment).\n\nI am preparing a follow-up patch to align test fixtures with current types and to make the signal-listener mock type-safe.

File: src/agents/execution-health.ts [hunk 1]
```diff
@@ -0,0 +1,407 @@
...
+
+  // -------------------------------------------------------------------------
+  // Pattern detectors
+  // -------------------------------------------------------------------------
+
+  private detectFileBurst(toolCalls: ToolCallEntry[]): ExecutionHealthSignal | undefined {
+    const writes = toolCalls.filter((tc) => tc.isWrite);
+    if (writes.length < this.cfg.fileBurstThreshold) {
+      return undefined;
+    }
```
[the remark this answers is not in the document]
Comment (dsantoreis) [reply src/agents/execution-health.ts]: The delta-based approach is by design. File burst detection looks at the current turn's tool calls to catch rapid-fire file creation within a single turn (the primary death spiral pattern). Cross-turn accumulation is handled by the error cascade detector. Merging both into one detector would conflate different failure modes.

File: src/config/types.agent-defaults.ts [hunk 2]
```diff
@@ -284,6 +284,25 @@ export type AgentDefaultsConfig = {
...
+  fileBurstWindowMs?: number;
+  /** Same tool+args repeats before warning (default: 5). */
+  toolRepeatThreshold?: number;
+  /** Window in ms for tool repeat detection (default: 300000). */
+  toolRepeatWindowMs?: number;
+  /** Consecutive turns without a real side-effect before warning (default: 10). */
+  noEffectLoopThreshold?: number;
+  /** Consecutive tool errors before warning (default: 3). */
+  errorCascadeThreshold?: number;
 };
```
[the remark this answers is not in the document]
Comment (dsantoreis) [reply src/config/types.agent-defaults.ts]: Valid. The type should be imported from execution-health.ts rather than redefined. Structural compatibility works today but this is fragile. Will fix if this PR moves forward.

File: src/agents/execution-health.ts [hunk 3]
```diff
@@ -0,0 +1,407 @@
...
+    };
+  }
+
+  private detectToolRepeat(
+    newCalls: ToolCallEntry[],
+    messages: AgentMessage[],
+    prePromptMessageCount: number,
+  ): ExecutionHealthSignal | undefined {
+    // Collect all tool calls in the session (not just new ones) for repeat detection
+    const allCalls = extractToolCalls(messages, prePromptMessageCount);
```
[the remark this answers is not in the document]
Comment (dsantoreis) [reply src/agents/execution-health.ts]: The full scan is intentional for the repeat detector. Tool repetition is a session-level pattern (same tool called 50+ times across the session), not a per-turn metric. The O(n) cost is bounded by session length which is bounded by compaction. For the typical session size (< 200 messages), this is negligible.

File: src/agents/execution-health.ts [hunk 4]
```diff
@@ -0,0 +1,407 @@
...
+
+export type ExecutionHealthConfig = {
+  enabled?: boolean;
+  fileBurstThreshold?: number;
+  fileBurstWindowMs?: number;
+  toolRepeatThreshold?: number;
+  toolRepeatWindowMs?: number;
+  noEffectLoopThreshold?: number;
+  errorCascadeThreshold?: number;
+};
```
[the remark this answers is not in the document]
Comment (dsantoreis) [reply src/agents/execution-health.ts:47]: Correct. The windowMs fields are config placeholders for future time-windowed filtering. They're exposed in config for forward compatibility but not yet wired into the detection logic. The current implementation uses message-count windows instead. I should document this gap or wire them in.

File: src/agents/execution-health.ts [hunk 5]
```diff
@@ -0,0 +1,407 @@
...
+    messages: AgentMessage[];
+    prePromptMessageCount: number;
+  }): ExecutionHealthSignal[] {
+    if (!this.cfg.enabled) {
+      return [];
+    }
+
+    const { messages, prePromptMessageCount } = params;
+    const startIndex = Math.max(prePromptMessageCount, this.lastEvaluatedIndex);
+    const toolCalls = extractToolCalls(messages, startIndex);
```
[the remark this answers is not in the document]
Comment (dsantoreis) [reply src/agents/execution-health.ts:251]: See response to the greptile comment above. Per-turn detection is intentional. Cross-turn patterns (creating 3 files per turn for 10 turns) are caught by the tool repeat detector, not the burst detector. Each detector covers a different failure mode.

File: src/agents/execution-health.ts [hunk 6]
```diff
@@ -0,0 +1,407 @@
...
+  }
+
+  private detectToolRepeat(
+    newCalls: ToolCallEntry[],
+    messages: AgentMessage[],
+    prePromptMessageCount: number,
+  ): ExecutionHealthSignal | undefined {
+    // Collect all tool calls in the session (not just new ones) for repeat detection
+    const allCalls = extractToolCalls(messages, prePromptMessageCount);
+    const counts = new Map<string, number>();
```
[the remark this answers is not in the document]
Comment (dsantoreis) [reply src/agents/execution-health.ts]: The tool repeat detector does scan from session start. The prePromptMessageCount parameter sets the boundary for the current turn's delta, but extractToolCalls for the repeat path processes all messages. The naming could be clearer.

File: src/agents/execution-health.ts [hunk 7]
```diff
@@ -0,0 +1,407 @@
...
+  }
+
+  private detectErrorCascade(
+    _newCalls: ToolCallEntry[],
+    messages: AgentMessage[],
+    prePromptMessageCount: number,
+  ): ExecutionHealthSignal | undefined {
+    // Walk backwards from the end to count consecutive errors
+    let consecutiveErrors = 0;
+    for (let i = messages.length - 1; i >= prePromptMessageCount; i--) {
```
[the remark this answers is not in the document]
Comment (dsantoreis) [reply src/agents/execution-health.ts:414]: Error cascade detection walks backward from the latest message regardless of turn boundaries. The prePromptMessageCount is used as a starting point for the scan, not a hard stop. The backward walk continues until it finds a non-error message or reaches the beginning.

File: src/agents/execution-health.ts [hunk 8]
```diff
@@ -0,0 +1,407 @@
...
+    prePromptMessageCount: number;
+  }): ExecutionHealthSignal[] {
+    if (!this.cfg.enabled) {
+      return [];
+    }
+
+    const { messages, prePromptMessageCount } = params;
+    const startIndex = Math.max(prePromptMessageCount, this.lastEvaluatedIndex);
+    const toolCalls = extractToolCalls(messages, startIndex);
+    this.lastEvaluatedIndex = messages.length;
```
[the remark this answers is not in the document]
Comment (dsantoreis) [reply src/agents/execution-health.ts:252]: Good edge case. After compaction shrinks the transcript, prePromptMessageCount could exceed the actual message array length. The extractToolCalls function handles this safely because it iterates from startIndex with bounds checking, but an explicit clamp would make the intent clearer. Will add a Math.min guard if this PR moves forward.

File: src/agents/execution-health.ts [hunk 9]
```diff
@@ -0,0 +1,407 @@
...
+      >;
+      const hasToolResult = content.some((b) => b.type === "tool_result");
+      if (!hasToolResult) {
+        continue;
+      }
+
+      const allErrors = content.filter((b) => b.type === "tool_result").every((b) => b.is_error);
+
+      if (allErrors) {
+        consecutiveErrors++;
```
[the remark this answers is not in the document]
Comment (dsantoreis) [reply src/agents/execution-health.ts]: The per-user-message counting is a conservative simplification. A single user message can contain multiple tool_result blocks, but counting per-message avoids double-counting when one tool failure triggers retry logic that produces multiple error results. Per-result counting could inflate the cascade score. Open to maintainer preference on granularity.

File: src/agents/pi-embedded-runner/run/attempt.ts [hunk 10]
```diff
@@ -754,6 +756,10 @@ export async function runEmbeddedAttempt(
     `embedded run start: runId=${params.runId} sessionId=${params.sessionId} provider=${params.provider} model=${params.modelId} thinking=${params.thinkLevel} messageChannel=${params.messageChannel ?? params.messageProvider ?? "unknown"}`,
   );
 
+  const executionHealthMonitor = new ExecutionHealthMonitor(
+    params.config?.agents?.defaults?.executionHealth,
+  );
```
[the remark this answers is not in the document]
Comment (dsantoreis) [reply src/agents/pi-embedded-runner/run/attempt.ts:1386]: ExecutionHealthMonitor is intentionally stateless across attempts. Each attempt gets a fresh evaluation because compaction between attempts can change the message history. Persisting state across attempts would require reconciling with post-compaction transcript changes, which adds complexity without clear benefit since the detector re-evaluates the full available history each time.

File: src/agents/execution-health.ts [hunk 11]
```diff
@@ -0,0 +1,407 @@
...
+    const msg = messages[i];
+    if (msg.role !== "assistant") {
+      continue;
+    }
+
+    const content = Array.isArray(msg.content) ? msg.content : [];
+    for (const rawBlock of content) {
+      // Content blocks arrive as Anthropic API shapes at runtime; cast to generic record.
+      const block = rawBlock as unknown as Record<string, unknown>;
+      if (block.type !== "tool_use") {
```
[the remark this answers is not in the document]
Comment (dsantoreis) [reply src/agents/execution-health.ts]: The extractToolCalls function handles both the raw Anthropic format (`type: tool_use` with `.input`) and the normalized format. The canonical transcript stores messages in the provider's native format before normalization. If the runner normalizes before storing, this is a valid concern. Will verify the actual transcript format in the runner integration.

File: src/agents/execution-health.ts [hunk 12]
```diff
@@ -0,0 +1,407 @@
...
+  private detectErrorCascade(
+    _newCalls: ToolCallEntry[],
+    messages: AgentMessage[],
+    prePromptMessageCount: number,
+  ): ExecutionHealthSignal | undefined {
+    // Walk backwards from the end to count consecutive errors
+    let consecutiveErrors = 0;
+    for (let i = messages.length - 1; i >= prePromptMessageCount; i--) {
+      const msg = messages[i];
+      if (msg.role !== "user") {
```
[the remark this answers is not in the document]
Comment (dsantoreis) [reply src/agents/execution-health.ts]: Duplicate of comment 9 about per-tool-result vs per-message counting. Same response applies: per-message counting is a conservative simplification to avoid inflation from retry-driven multiple error results within one turn.

File: src/agents/execution-health.ts [hunk 13]
```diff
@@ -0,0 +1,407 @@
...
+    const content = Array.isArray(msg.content) ? msg.content : [];
+    for (const rawBlock of content) {
+      // Content blocks arrive as Anthropic API shapes at runtime; cast to generic record.
+      const block = rawBlock as unknown as Record<string, unknown>;
+      if (block.type !== "tool_use") {
+        continue;
+      }
+
+      const name = block.name as string;
+      const args = block.input;
```
[the remark this answers is not in the document]
Comment (dsantoreis) [reply src/agents/execution-health.ts]: Good catch. The `extractToolCalls` function is intentionally scoped to the Anthropic wire format (`tool_use` with `.input`) because the canonical transcript stores messages in the provider's native shape. The normalized formats (`toolCall`/`toolUse`/`functionCall`) appear in the runner's internal processing pipeline but are not persisted to the session JSONL that `ExecutionHealthMonitor` reads.

That said, if OpenClaw starts persisting normalized formats in the future (or if other providers store different shapes), this will break silently. Adding a comment documenting this assumption.

File: src/agents/execution-health.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (dsantoreis) [reply src/agents/execution-health.ts]: Acknowledged. The `windowMs` fields are config placeholders for future time-windowed filtering, as noted in the earlier response thread. Current detection is per-turn (bounded by `prePromptMessageCount`), which is the primary death spiral signal. Wiring up timestamp-based windowing is planned but not blocking for the initial implementation since per-turn burst detection catches the acute pattern.

File: src/agents/execution-health.ts [hunk 14]
```diff
@@ -0,0 +1,407 @@
...
+  private detectToolRepeat(
+    newCalls: ToolCallEntry[],
+    messages: AgentMessage[],
+    prePromptMessageCount: number,
+  ): ExecutionHealthSignal | undefined {
+    // Collect all tool calls in the session (not just new ones) for repeat detection
+    const allCalls = extractToolCalls(messages, prePromptMessageCount);
+    const counts = new Map<string, number>();
+
+    for (const tc of allCalls) {
```
[the remark this answers is not in the document]
Comment (dsantoreis) [reply src/agents/execution-health.ts]: Same reasoning as the burst window comment. The repeat detector intentionally scans the full session because death spiral tool repetition is a session-level pattern (e.g., the same `write` call 50+ times across many turns). Time windowing would miss slow-burn spirals. The `windowMs` config field is reserved for a future refinement where both absolute count and time-bounded rate can be evaluated together.

File: src/agents/execution-health.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (dsantoreis) [reply src/agents/execution-health.ts]: Good catch, fixed in 2a75190. Execution health now accepts raw Anthropic tool_use blocks and normalized toolCall/toolUse/functionCall shapes, so the monitor sees the assistant tool calls before evaluating health.

File: src/agents/execution-health.ts [hunk 15]
```diff
@@ -0,0 +1,423 @@
...
+    _newCalls: ToolCallEntry[],
+    messages: AgentMessage[],
+    prePromptMessageCount: number,
+  ): ExecutionHealthSignal | undefined {
+    // Walk backwards from the end to count consecutive errors
+    let consecutiveErrors = 0;
+    for (let i = messages.length - 1; i >= prePromptMessageCount; i--) {
+      const msg = messages[i];
+      if (msg.role !== "user") {
+        continue;
```
[the remark this answers is not in the document]
Comment (dsantoreis) [reply src/agents/execution-health.ts]: Good catch, fixed in c7d8e1fce. Error cascade detection now counts SDK-persisted toolResult turns in addition to Anthropic user tool_result batches, and I added a regression test that covers three consecutive SDK toolResult errors.

File: src/config/types.agent-defaults.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (dsantoreis) [reply src/config/types.agent-defaults.ts]: Good catch, fixed in e1ab51d80. ExecutionHealthConfig now has a single source of truth in src/agents/execution-health.ts and config types import it instead of duplicating the shape.

File: src/agents/execution-health.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (dsantoreis) [reply src/agents/execution-health.ts]: Good catch, documented in 2c35e0844. execution-health now explicitly notes that it reads provider-native persisted transcript shapes today, and that normalized runner-only tool blocks are not expected in session JSONL history yet.

File: src/agents/execution-health.ts [hunk 16]
```diff
@@ -0,0 +1,437 @@
...
+        fileCreations: allCalls.filter((tc) => tc.isWrite).length,
+        repeatedTools: [...new Set(repeated)],
+      },
+      recommendation: `Tools repeated ≥${this.cfg.toolRepeatThreshold} times with identical args: ${[...new Set(repeated)].join(", ")}. The agent may be stuck in a loop.`,
+    };
+  }
+
+  private detectNoEffectLoop(toolCalls: ToolCallEntry[]): ExecutionHealthSignal | undefined {
+    const hasEffect = toolCalls.some((tc) => tc.isEffect);
+    if (hasEffect) {
```
[the remark this answers is not in the document]
Comment (dsantoreis) [reply src/agents/execution-health.ts]: Leaving this for maintainer input on whether no-effect-loop should reset only on successful side effects or on attempted side effects too.

File: src/agents/execution-health.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (dsantoreis) [reply src/agents/execution-health.ts]: Documented in e92a068b4. The execution-health config now explicitly marks fileBurstWindowMs and toolRepeatWindowMs as reserved placeholders for future time-window filtering, while the current implementation remains turn-bounded for bursts and session-wide for repeats.

File: src/agents/execution-health.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (dsantoreis) [reply src/agents/execution-health.ts]: Documented in e92a068b4. The execution-health config now explicitly marks fileBurstWindowMs and toolRepeatWindowMs as reserved placeholders for future time-window filtering, while the current implementation remains turn-bounded for bursts and session-wide for repeats.

File: src/agents/execution-health.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (dsantoreis) [reply src/agents/execution-health.ts]: Good catch, fixed in f8432a313. no-effect-loop now only resets on successful effect calls, so failed git push/send attempts keep contributing to the streak, and I added a regression test that covers a failing git push tool_result.

File: src/agents/execution-health.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (dsantoreis) [reply src/agents/execution-health.ts:252]: Good catch, fixed in 88b33d191. execution-health now clamps both prePromptMessageCount and the stored lastEvaluatedIndex to the current transcript length after compaction, and I added regression coverage for the shrunken-history path.

File: src/agents/execution-health.ts [hunk 17]
```diff
@@ -0,0 +1,439 @@
...
+        continue;
+      }
+
+      const name = toolCall.name ?? "unknown";
+      const args = block.input ?? block.arguments ?? block.args ?? {};
+
+      let isError = false;
+      for (let j = i + 1; j < messages.length && j <= i + 3; j++) {
+        if (isToolResultErrorMessage(messages[j], toolCall.id)) {
+          isError = true;
```
[the remark this answers is not in the document]
Comment (dsantoreis) [reply src/agents/execution-health.ts:177]: Good catch, fixed in 5ad3f946. We now scan the full following user tool-result turn until the next assistant message, so later results in the same batch still mark failed effect calls as errors. Added regression coverage for the out-of-order and later-result case too.

File: src/agents/execution-health.ts [hunk 18]
```diff
@@ -0,0 +1,469 @@
...
+
+// Tools that are expected to be called repeatedly with the same args.
+const REPEAT_IGNORE_TOOLS = new Set(["Read", "read", "memory_search", "MemorySearch"]);
+
+// Tool names / shell commands that count as "real effects".
+const EFFECT_TOOL_NAMES = new Set([
+  "Bash",
+  "bash",
+  "computer",
+  "execute_command",
```
[the remark this answers is not in the document]
Comment (dsantoreis) [reply src/agents/execution-health.ts:71]: Good catch, fixed in a3a1427bb. exec now counts as an effect tool, so successful exec-based git flows reset the no-effect streak correctly.

File: src/agents/execution-health.ts [hunk 19]
```diff
@@ -0,0 +1,469 @@
...
+      }
+
+      const argsObj = (args && typeof args === "object" ? args : {}) as Record<string, unknown>;
+
+      const isWrite =
+        (name === "Write" ||
+          name === "write" ||
+          name === "write_to_file" ||
+          name === "create_file") &&
+        typeof argsObj.file_path === "string";
```
[the remark this answers is not in the document]
Comment (dsantoreis) [reply src/agents/execution-health.ts:189]: Good catch, fixed in a3a1427bb. write detection now accepts both path and file_path, matching the normalized tool args we persist.

File: src/agents/execution-health.ts [hunk 20]
```diff
@@ -0,0 +1,473 @@
...
+    prePromptMessageCount: number;
+  }): ExecutionHealthSignal[] {
+    if (!this.cfg.enabled) {
+      return [];
+    }
+
+    const { messages, prePromptMessageCount } = params;
+    const boundedPrePromptCount = Math.min(prePromptMessageCount, messages.length);
+    const previousIndex = Math.min(this.lastEvaluatedIndex, messages.length);
+    const startIndex = Math.max(boundedPrePromptCount, previousIndex);
```
[the remark this answers is not in the document]
Comment (dsantoreis) [reply src/agents/execution-health.ts]: Good catch, fixed in a3a1427bb. The monitor now fingerprints the evaluated transcript prefix and rescans safely after compaction, so shifted post-compaction turns do not get skipped.

File: src/agents/execution-health.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (dsantoreis) [reply src/agents/execution-health.ts:414]: Good catch, fixed in e748c95b8. error-cascade now scans backward across the whole available transcript instead of stopping at the current turn boundary, and I added regression coverage for repeated single-error turns.

File: src/agents/execution-health.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (dsantoreis) [reply src/agents/execution-health.ts]: Good catch, locked down with coverage in 0b48c91c6. execution-health already uses extractToolCallsFromAssistant, which accepts toolCall, toolUse, functionCall, and tool_use blocks. I added a regression test that exercises normalized assistant tool-call shapes so this path stays covered.

File: src/agents/execution-health.ts (hunk 18 above)
[the remark this answers is not in the document]
Comment (dsantoreis) [reply src/agents/execution-health.ts:71]: Good catch, fixed in a3a1427. exec is now included in effect detection, and 0b48c91 adds coverage for the normalized exec path.

File: src/agents/execution-health.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (dsantoreis) [reply src/agents/execution-health.ts]: Good catch, fixed in e92a068. No-effect now resets only on successful effect calls, so failed git push style loops still count toward the streak.

File: src/agents/execution-health.ts (hunk 17 above)
[the remark this answers is not in the document]
Comment (dsantoreis) [reply src/agents/execution-health.ts:177]: Good catch, fixed in 5ad3f94. We now scan the full tool-result portion of the turn instead of stopping after three messages.

File: src/agents/execution-health.ts (hunk 19 above)
[the remark this answers is not in the document]
Comment (dsantoreis) [reply src/agents/execution-health.ts:189]: Good catch, fixed in a3a1427. Write detection now accepts both file_path and the normalized path field.

File: src/agents/execution-health.ts (hunk 20 above)
[the remark this answers is not in the document]
Comment (dsantoreis) [reply src/agents/execution-health.ts]: Good catch, fixed in a3a1427. When compaction shrinks history we now restart the scan from 0 for that evaluation, so the current turn is not skipped.

**Comment by dsantoreis** [comment]
Follow-up pushed in `906e876ea`.

- accumulate file-burst detection across turns inside the configured window instead of only counting writes from the latest delta
- switch repeat hashing to a bounded serializer so large tool args do not force unbounded `JSON.stringify` work
- stop re-scanning the full transcript on every evaluate by keeping running repeat counts
- add a regression test that reproduces cross-turn file bursts

Validation: `pnpm test src/agents/execution-health.test.ts`

File: src/agents/execution-health.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (dsantoreis) [reply src/agents/execution-health.ts]: Good catch, fixed in e748c95b8. error-cascade now walks backward across the persisted transcript and counts consecutive failing toolResult messages across turn boundaries, so repeated single-error turns trigger the warning again. Added regression coverage for that path too.

File: src/agents/execution-health.ts [hunk 21]
```diff
@@ -0,0 +1,541 @@
...
+    if (seen.has(current)) {
+      budget -= 7;
+      return "<cycle>";
+    }
+    seen.add(current);
+    if (Array.isArray(current)) {
+      return current.slice(0, 25).map((entry) => visit(entry, depth + 1));
+    }
+    const out: Record<string, unknown> = {};
+    for (const [key, entry] of Object.entries(current).slice(0, 25)) {
```
[the remark this answers is not in the document]
Comment (dsantoreis) [reply src/agents/execution-health.ts]: Good catch. Leaving this for maintainer input on whether we should canonicalize arg keys in this hot path or gate it behind a threshold to avoid extra overhead.

File: src/agents/execution-health.ts [hunk 22]
```diff
@@ -0,0 +1,548 @@
...
+
+      let isEffect = false;
+      if (EFFECT_TOOL_NAMES.has(name)) {
+        const cmd = typeof argsObj.command === "string" ? argsObj.command : "";
+        if (cmd) {
+          isEffect = EFFECT_COMMAND_PATTERNS.some((re) => re.test(cmd));
+        }
+      }
+      if (name.includes("send") || name.includes("Send") || name.includes("message")) {
+        isEffect = true;
```
[the remark this answers is not in the document]
Comment (dsantoreis) [reply src/agents/execution-health.ts:199]: Good catch, fixed in 7df7cd84e. no-effect now treats message read/search/reactions/list-style actions as non-mutating, so only mutating message actions reset the streak. Added regression coverage in execution-health.test.ts as well.

File: src/agents/execution-health.ts [hunk 23]
```diff
@@ -0,0 +1,553 @@
...
+  "bash",
+  "computer",
+  "exec",
+  "execute_command",
+  "run_terminal_command",
+]);
+
+const EFFECT_COMMAND_PATTERNS = [
+  /\bgit\s+(commit|push|merge|rebase)\b/,
+  /\bgh\s+(pr|issue)\b/,
```
[the remark this answers is not in the document]
Comment (dsantoreis) [reply src/agents/execution-health.ts:77]: Good catch, fixed in d218ee42. We now only treat mutating gh subcommands as effect signals, so read-only gh calls no longer reset no-effect streaks.

File: src/agents/execution-health.ts [hunk 24]
```diff
@@ -0,0 +1,553 @@
...
+  }): ExecutionHealthSignal[] {
+    if (!this.cfg.enabled) {
+      return [];
+    }
+
+    const { messages, prePromptMessageCount } = params;
+    const sharedPrefixLength = Math.min(this.lastEvaluatedIndex, messages.length);
+    const transcriptMutated = Array.from({ length: sharedPrefixLength }).some((_, index) => {
+      return this.lastTranscriptFingerprints[index] !== fingerprintMessage(messages[index]);
+    });
```
[the remark this answers is not in the document]
Comment (dsantoreis) [reply src/agents/execution-health.ts]: Good catch, fixed in 187083b5. We now hash and compare transcript fingerprints incrementally instead of re-hashing the full shared prefix on every evaluate call.

File: src/agents/execution-health.ts [hunk 25]
```diff
@@ -0,0 +1,553 @@
...
+  }
+
+  return entries;
+}
+
+// ---------------------------------------------------------------------------
+// Monitor
+// ---------------------------------------------------------------------------
+
+export class ExecutionHealthMonitor {
```
[the remark this answers is not in the document]
Comment (dsantoreis) [reply src/agents/execution-health.ts:221]: This one is stale. The monitor is already wired in runtime attempt flow and executes during embedded attempts. Latest branch head is 2b789ea and CI is green.

**Comment by dsantoreis** [comment]
Addressed the security comment and pushed follow-up commit `8fbcdf263`.

Changes:
- removed raw `args` retention from `ToolCallEntry`
- replaced pseudo-hash (`name:serializedArgs`) with SHA-256 digest over bounded serialized payload

Validation:
- `pnpm -C openclaw test src/agents/execution-health.test.ts`

File: src/agents/tool-call-id.ts [hunk 26]
```diff
@@ -4,7 +4,7 @@ import type { AgentMessage } from "@mariozechner/pi-agent-core";
 export type ToolCallIdMode = "strict" | "strict9";
 
 const STRICT9_LEN = 9;
-const TOOL_CALL_TYPES = new Set(["toolCall", "toolUse", "functionCall"]);
+const TOOL_CALL_TYPES = new Set(["toolCall", "toolUse", "functionCall", "tool_use"]);
```
[the remark this answers is not in the document]
Comment (dsantoreis) [reply src/agents/tool-call-id.ts:7]: Good callout. Leaving this for maintainer input on one point: this guard currently expects normalized `toolResult` role messages in the persisted transcript path, and I want to confirm whether Anthropic-style `user` + `tool_result` blocks can still reach this layer before we broaden `extractToolCallsFromAssistant` here.

[the remark this answers is not in the document]
Comment (dsantoreis) [reply src/agents/tool-call-id.ts:7]: Good callout. Leaving this for maintainer input on whether we should gate tool_use parsing behind a consumer-specific path instead of broadening TOOL_CALL_TYPES globally.

File: src/agents/pi-embedded-runner/run/attempt.ts [hunk 27]
```diff
@@ -413,14 +415,12 @@ export function wrapOllamaCompatNumCtx(baseFn: StreamFn | undefined, numCtx: num
       ...options,
       onPayload: (payload: unknown) => {
         if (!payload || typeof payload !== "object") {
-          return options?.onPayload?.(payload, model);
         }
         const payloadRecord = payload as Record<string, unknown>;
         if (!payloadRecord.options || typeof payloadRecord.options !== "object") {
```
[the remark this answers is not in the document]
Comment (dsantoreis) [reply src/agents/pi-embedded-runner/run/attempt.ts:420]: Good catch, this is a real regression risk. I will update the wrapper to preserve the existing onPayload chain and harden non-object payload handling, then push a follow-up commit.
